### PR TITLE
Contracts and Harnesses for `<*const T>::add`, `sub` and `offset`

### PR DESCRIPTION
**Pull request by xsxszab**
Towards #76 

**Summary**
* Adds contracts for <*const T>::add, sub and offset.
* Adds proof for contracts for above methods, verifying following pointee types:
    * All integer types
    * Tuples (representing composite types)
    * Slices
    * Unit type

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by stogaru** [comment]
@feliperodri FYI

File: library/core/src/ptr/const_ptr.rs [hunk 1]
```diff
@@ -1899,3 +1947,219 @@ impl<T: ?Sized> PartialOrd for *const T {
...
+    generate_slice_harnesses!(i128, check_const_offset_slice_i128, check_const_add_slice_i128, check_const_sub_slice_i128);
+    generate_slice_harnesses!(isize, check_const_offset_slice_isize, check_const_add_slice_isize, check_const_sub_slice_isize);
+    generate_slice_harnesses!(u8, check_const_offset_slice_u8, check_const_add_slice_u8, check_const_sub_slice_u8);
+    generate_slice_harnesses!(u16, check_const_offset_slice_u16, check_const_add_slice_u16, check_const_sub_slice_u16);
+    generate_slice_harnesses!(u32, check_const_offset_slice_u32, check_const_add_slice_u32, check_const_sub_slice_u32);
+    generate_slice_harnesses!(u64, check_const_offset_slice_u64, check_const_add_slice_u64, check_const_sub_slice_u64);
+    generate_slice_harnesses!(u128, check_const_offset_slice_u128, check_const_add_slice_u128, check_const_sub_slice_u128);
+    generate_slice_harnesses!(usize, check_const_offset_slice_usize, check_const_add_slice_usize, check_const_sub_slice_usize);
+
+    // Generate slice harnesses for tuples (offset, add, sub)
```
Comment (Eh2406) [inline library/core/src/ptr/const_ptr.rs:2025]: Do we need the `()` case for `generate_slice_harnesses`, the way we have it for `generate_const_arithmetic_harness`?

Comment (szlee118) [reply library/core/src/ptr/const_ptr.rs:2025]: Including it in the slice harnesses ensures that our verification covers all possible types, and the overhead should be minimal.

File: library/core/src/ptr/const_ptr.rs [hunk 2]
```diff
@@ -1899,3 +1947,219 @@ impl<T: ?Sized> PartialOrd for *const T {
...
+
+    // fn <*const T>::offset() unit type verification
+    generate_const_arithmetic_harness!((), check_const_offset_unit, offset);
+
+    // fn <*const T>::offset() composite type verification
+    generate_const_arithmetic_harness!((i8, i8), check_const_offset_tuple_1, offset);
+    generate_const_arithmetic_harness!((f64, bool), check_const_offset_tuple_2, offset);
+    generate_const_arithmetic_harness!((i32, f64, bool), check_const_offset_tuple_3, offset);
+    generate_const_arithmetic_harness!((i8, u16, i32, u64, isize), check_const_offset_tuple_4, offset);
+}
```
Comment (feliperodri) [inline library/core/src/ptr/const_ptr.rs:2165]: New line.

File: library/core/src/ptr/const_ptr.rs [hunk 3]
```diff
@@ -1899,3 +1947,219 @@ impl<T: ?Sized> PartialOrd for *const T {
         *self >= *other
     }
 }
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+mod verify {
+    use crate::kani;
+    use core::mem;
+    // Constant for array size used in all tests, for performance reason
```
Comment (feliperodri) [inline library/core/src/ptr/const_ptr.rs:1956]: Could you give a bit more info? You're using a `BUF_SIZE` of 200 for other harnesses. How is the verification time for 5 here? Does it blow up if you double it? Is this enough to explore all functions for slices?

**Review by feliperodri — COMMENTED** [review]
I'll approve it after we merge https://github.com/model-checking/verify-rust-std/pull/113

File: library/core/src/ptr/const_ptr.rs (hunk 3 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs:1956]: Hi @szlee118, could you elaborate on this matter?

File: library/core/src/ptr/const_ptr.rs (hunk 2 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs]: Resolved in commit 5edd16e and dd2a8eb.

**Comment by feliperodri** [comment]
@xsxszab could you solve the conflicts?

File: library/core/src/ptr/const_ptr.rs (hunk 3 above)
Comment (szlee118) [reply library/core/src/ptr/const_ptr.rs:1956]: The verification time slightly increase, but it doesn't blow up when doubling it.

`BUF_SIZE`:
- Represents the memory buffer size (in bytes) allocated for testing raw pointer arithmetic.
- Needs to be large enough to ensure pointer operations like add, sub, and offset have enough valid memory space to operate on, covering the full range of test scenarios.

`ARRAY_SIZE`:
- Represents the number of elements in the slice or array being tested.
- Doesn't need to be large because all possible values for the slice are explored, including edge cases like pointers at the start, middle, and end of the slice. Symbolic execution generalizes across all possible elements, regardless of the actual array size.

**Comment by szlee118** [comment]
@feliperodri Conflicts resolved, and I removed the comment stating `performance reason` improperly.

File: library/core/src/ptr/const_ptr.rs [hunk 4]
```diff
@@ -409,6 +409,20 @@ impl<T: ?Sized> *const T {
     #[rustc_const_stable(feature = "const_ptr_offset", since = "1.61.0")]
     #[inline(always)]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // Precondition 1: the computed offset `count * size_of::<T>()` does not overflow `isize`
+        count.checked_mul(core::mem::size_of::<T>() as isize).is_some() &&
+        // Precondition 2: adding the computed offset to `self` does not cause overflow
+        (self as isize).checked_add((count * core::mem::size_of::<T>() as isize)).is_some() &&
```
Comment (feliperodri) [inline library/core/src/ptr/const_ptr.rs:416]: There is an implicit condition here. What happens if `core::mem::size_of::<T>()` represents a *very* larger type in a 64-bit system? Wouldn't the conversions to `isize` resolve this to checking if the wrong negative number fits the specification? I'm a bit curious on if you gave that a thought. For the types that you instantiated in the harnesses, this will not be a problem at all. This is not a blocker.

File: library/core/src/ptr/const_ptr.rs [hunk 5]
```diff
@@ -1920,6 +1964,279 @@ mod verify {
     use crate::kani;
     use core::mem;
     use kani::PointerGenerator;
+    // Constant for array size used in all tests
+    const ARRAY_SIZE: usize = 5;
```
Comment (feliperodri) [inline library/core/src/ptr/const_ptr.rs:1968]: Could you add a proper justification for `5`? How this "bound" impacts the proofs? What is the verification performance for this number? Are we missing anything by staying within this bound?

File: library/core/src/ptr/const_ptr.rs [hunk 6]
```diff
@@ -1920,6 +1964,279 @@ mod verify {
...
+                let offset: usize = kani::any();
+                let count: usize = kani::any();
+                kani::assume(offset <= ARRAY_SIZE * mem::size_of::<$ty>());
+                let ptr_with_offset: *const $ty = test_ptr.wrapping_byte_add(offset);
+                unsafe {
+                    ptr_with_offset.sub(count);
+                }
+            }
+        };
+    }
```
Comment (feliperodri) [inline library/core/src/ptr/const_ptr.rs:2020]: You can reduce this code by 3x if you pass the operation under verification (i.e., offset, add, or sub) to the macro.

File: library/core/src/ptr/const_ptr.rs [hunk 7]
```diff
@@ -1920,6 +1964,279 @@ mod verify {
...
+                const BUF_SIZE: usize = 200;
+                let mut generator = kani::PointerGenerator::<BUF_SIZE>::new();
+                let test_ptr: *const $type = generator.any_in_bounds().ptr;
+                let count: isize = kani::any();
+                unsafe {
+                    test_ptr.offset(count);
+                }
+            }
+        };
+    }
```
Comment (feliperodri) [inline library/core/src/ptr/const_ptr.rs:2166]: you can reduce 3x this macro passing the operation to the macro.

File: library/core/src/ptr/const_ptr.rs [hunk 8]
```diff
@@ -2038,6 +2355,7 @@ mod verify {
         check_const_offset_from_isize_arr
     );
 
+    // fn <*const T>::offset_from() typle and tuple slice types verification
```
Comment (feliperodri) [inline library/core/src/ptr/const_ptr.rs:2358]: typo

File: library/core/src/ptr/const_ptr.rs (hunk 4 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs:416]: Based on the information in the Rust documentation [here](https://doc.rust-lang.org/stable/reference/types/numeric.html#machine-dependent-integer-types), the size of `T` cannot exceed `isize::MAX` on any platform. As a result, the conversion from `usize` to `isize` will not cause an overflow, ensuring that `core::mem::size_of::<T>() as isize` will always be non-negative.

File: library/core/src/ptr/const_ptr.rs (hunk 5 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs:1968]: @szlee118 FYI

File: library/core/src/ptr/const_ptr.rs (hunk 6 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs]: Addressed in new commits.

File: library/core/src/ptr/const_ptr.rs (hunk 7 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs]: Addressed in new commits.

File: library/core/src/ptr/const_ptr.rs (hunk 8 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs]: The typo has been fixed

File: library/core/src/ptr/const_ptr.rs (hunk 5 above)
Comment (szlee118) [reply library/core/src/ptr/const_ptr.rs:1969]: There is an [explanation](https://github.com/model-checking/verify-rust-std/pull/166#discussion_r1861482793) in previous comments. Just to confirm if this justification is required to be comment in the code? We are currently gathering verfication details in final documents.

File: library/core/src/ptr/const_ptr.rs (hunk 3 above)
Comment (tautschnig) [reply library/core/src/ptr/const_ptr.rs]: I'd much appreciate if the above comment could be reflected as a comment in the source code. Anyone looking at those constants in future should know why those particular numbers have been chosen.

File: library/core/src/ptr/const_ptr.rs (hunk 5 above)
Comment (tautschnig) [reply library/core/src/ptr/const_ptr.rs:1969]: See https://github.com/model-checking/verify-rust-std/pull/166#discussion_r1864993769 for my 2c on this.

File: library/core/src/ptr/const_ptr.rs (hunk 4 above)
Comment (tautschnig) [inline library/core/src/ptr/const_ptr.rs:416]: I believe you could introduce a local variable to hold the result of `count.checked_mul(core::mem::size_of::<T>() as isize)`. This would enable avoiding multiplying twice (which is a costly operation in Kani).

File: library/core/src/ptr/const_ptr.rs [hunk 9]
```diff
@@ -932,6 +946,21 @@ impl<T: ?Sized> *const T {
     #[rustc_const_stable(feature = "const_ptr_offset", since = "1.61.0")]
     #[inline(always)]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // Precondition 1: the computed offset `count * size_of::<T>()` does not overflow `isize`
+        count.checked_mul(core::mem::size_of::<T>()).is_some() &&
+        count * core::mem::size_of::<T>() <= isize::MAX as usize &&
+        // Precondition 2: adding the computed offset to `self` does not cause overflow
+        (self as isize).checked_add((count * core::mem::size_of::<T>()) as isize).is_some() &&
```
Comment (tautschnig) [inline library/core/src/ptr/const_ptr.rs:954]: As above: introduce a local variable to hold the result of the multiplication (which you are doing three times in this case).

File: library/core/src/ptr/const_ptr.rs [hunk 10]
```diff
@@ -1041,6 +1070,21 @@ impl<T: ?Sized> *const T {
     #[cfg_attr(bootstrap, rustc_allow_const_fn_unstable(unchecked_neg))]
     #[inline(always)]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // Precondition 1: the computed offset `count * size_of::<T>()` does not overflow `isize`
+        count.checked_mul(core::mem::size_of::<T>()).is_some() &&
+        count * core::mem::size_of::<T>() <= isize::MAX as usize &&
+        // Precondition 2: subtracting the computed offset from `self` does not cause overflow
+        (self as isize).checked_sub((count * core::mem::size_of::<T>()) as isize).is_some() &&
```
Comment (tautschnig) [inline library/core/src/ptr/const_ptr.rs:1078]: As above: local variable for the multiplication result.

**Review by tautschnig — APPROVED** [review]

File: library/core/src/ptr/const_ptr.rs (hunk 9 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs]: As above: addressed in the new commits.

File: library/core/src/ptr/const_ptr.rs (hunk 10 above)
Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs]: As above: addressed in the new commits.

File: library/core/src/ptr/const_ptr.rs (hunk 5 above)
Comment (feliperodri) [reply library/core/src/ptr/const_ptr.rs:1978]: Will this comment be address by another PR?

**Review by feliperodri — APPROVED** [review]

File: library/core/src/ptr/const_ptr.rs [hunk 11]
```diff
@@ -409,6 +409,23 @@ impl<T: ?Sized> *const T {
...
+        if let Some(computed_offset) = count.checked_mul(core::mem::size_of::<T>() as isize) {
+            // Precondition 2: adding the computed offset to `self` does not cause overflow.
+            (self as isize).checked_add(computed_offset).is_some()
+        } else {
+            false
+        } &&
+        // Precondition 3: If `T` is a unit type (`size_of::<T>() == 0`), this check is unnecessary as it has no allocated memory.
+        // Otherwise, for non-unit types, `self` and `self.wrapping_offset(count)` should point to the same allocated object,
+        // restricting `count` to prevent crossing allocation boundaries.
+        ((core::mem::size_of::<T>() == 0) || (kani::mem::same_allocation(self, self.wrapping_offset(count))))
```
Comment (celinval) [inline library/core/src/ptr/const_ptr.rs:423]: I recommend the following changes:
1. Start with the check that `T` is a ZST, in which case the operation should be safe.
2. You could also use `map_or(false, |computed_offset| {})` to replace the `if let` logic to make it a bit cleaner (not a big deal).
3.  Please don't use `kani` specific APIs in a contract. If you do need a tool specific API, the attribute should be conditional on the tool being enabled. I believe `same_allocation` has been added to `ub_checks` module. If it hasn't feel free to add it.

Something like:

```suggestion
        (core::mem::size_of::<T>() == 0)
        || (count.checked_mul(core::mem::size_of::<T>() as isize).map_or(false, |computed_offset| {
              (self as isize).checked_add(computed_offset).is_some()
            })
          && core::ub_checks::same_allocation(self, self.wrapping_offset(count)))
```

File: library/core/src/ptr/const_ptr.rs (hunk 5 above)
Comment (szlee118) [reply library/core/src/ptr/const_ptr.rs:1978]: Yes it will be addressed by another PR.

File: library/core/src/ptr/const_ptr.rs (hunk 11 above)
Comment (feliperodri) [reply library/core/src/ptr/const_ptr.rs:423]: Please, make sure to address this comment on the existing contracts added from https://github.com/model-checking/verify-rust-std/pull/113 and https://github.com/model-checking/verify-rust-std/pull/154.

Comment (feliperodri) [reply library/core/src/ptr/const_ptr.rs:423]: cc. @xsxszab @stogaru @szlee118 @unseenmars @MayureshJoshi25

Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs:423]: Since this PR is for verifying `const` functions, would it be appropriate to update the code for `mut` functions as well?

Comment (feliperodri) [reply library/core/src/ptr/const_ptr.rs:423]: you can address on a separate PR.

Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs:423]: Sure, I’ll submit a new PR once current comments are resolved.

Comment (xsxszab) [reply library/core/src/ptr/const_ptr.rs]: PR #203 has been created to refactor `mut` contracts and proof for contracts.

**Review by tautschnig — APPROVED** [review]
